File: apps/mobile/modules/podverse-media-engine/android/src/main/java/expo/modules/podversemediaengine/PodverseVideoSurfaceHost.kt
package expo.modules.podversemediaengine

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.view.TextureView
import android.view.View
import android.view.ViewGroup
import java.lang.ref.WeakReference

// PG-5 steps 2.15 + 2.17 (details 094, 096) + Plan 01 gap remediation (detail 099 addendum).
//
// The ONE native video surface for the process. Mini player and full player are two React views
// over a single ExoPlayer and a single TextureView; expanding **reparents** THIS view into the
// active RN target (2.20) instead of creating a second player or mounting a second RN <Video>
// (Track 11.18 anti-pattern).
//
// TextureView (not SurfaceView): SurfaceView has a separate compositor window and throws
// `IllegalStateException: child already has a parent` when reparented during React Navigation
// native-stack modal fragment transitions (MainActivity onPause). TextureView is a normal view and
// reparents cleanly between the mini and full PodverseVideoSurfaceView hosts.
object PodverseVideoSurfaceHost {
  /** Stable target ids the single surface can occupy. RN mounts a `PodverseVideoSurfaceView` each. */
  const val TARGET_MINI = "mini"
  const val TARGET_FULL = "full"

  private val mainHandler = Handler(Looper.getMainLooper())

  /** The one and only video TextureView. Bound to the shared ExoPlayer via the engine. */
  private var textureView: TextureView? = null

  /** RN-mounted target views by id. The single surface is reparented into the active target's view. */
  private val targetViews = mutableMapOf<String, WeakReference<ViewGroup>>()

  /** The target the surface currently occupies (`null` when unplaced / hidden). */
  private var activeTarget: String? = null

  /**
   * Last capability reported by the engine. The surface only shows when the current item actually
   * has video frames — so a video-medium item playing its audio enclosure never leaves a black
   * rectangle (2.23).
   */
  private var currentItemHasVideo = false

  /**
   * JS-desired visibility for the active target (RN drives this from the playback target kind via
   * `setVideoSurfaceVisible`). Final visibility = `desiredVisible && currentItemHasVideo`.
   */
  private var desiredVisible = false

  /**
   * Return the single [TextureView], creating it and binding it to the shared ExoPlayer on first
   * use. There is exactly one surface for the process; callers must not create their own.
   */
  fun getOrCreateSurface(context: Context): TextureView {
    return onMainSync {
      textureView
        ?: TextureView(context.applicationContext).also { view ->
          view.visibility = View.GONE
          textureView = view
          // Bind the one texture to the one shared player.
          PodverseAudioEngine.attachVideoTextureView(view)
          // Track video capability so the host can hide itself for audio-only items (policy 2.23).
          PodverseAudioEngine.onVideoCapabilityChanged = { hasVideo ->
            onMain {
              currentItemHasVideo = hasVideo
              updateVisibility()
            }
          }
        }
    }
  }

  // MARK: - RN-facing API (called by PodverseVideoSurfaceView + the Expo module)

  /**
   * Register (or replace) the RN view for [target]. The first registered target becomes active so
   * the single surface is placed; if [target] is active, reparent into it. Never loads/releases.
   */
  fun registerTargetView(target: String, view: ViewGroup) {
    onMain {
      targetViews[target] = WeakReference(view)
      if (activeTarget == null) {
        activeTarget = target
      }
      if (activeTarget == target) {
        reparentIntoActiveTarget()
      }
    }
  }

  /**
   * Unregister an RN view (on RN unmount). Only clears when the going-away view is the currently
   * registered one (avoids clobbering a re-registered view during a fast remount). If it was the
   * active target's view, detach the surface (kept alive; just no host).
   */
  fun unregisterTargetView(target: String, view: ViewGroup) {
    onMain {
      if (targetViews[target]?.get() === view) {
        targetViews.remove(target)
        if (activeTarget == target) {
          (textureView?.parent as? ViewGroup)?.removeView(textureView)
          updateVisibility()
        }
      }
    }
  }

  /**
   * Move the single surface to a registered target — bridge `animateVideoSurface` (2.19/2.20). Only
   * reparenting/geometry changes; the ExoPlayer and playhead are untouched.
   */
  fun setActiveTarget(target: String, animatedDurationMs: Long = 0) {
    onMain {
      activeTarget = target
      reparentIntoActiveTarget()
    }
  }

  /**
   * Set the JS-desired visibility for the active target (RN drives this from the playback target
   * kind). The surface only actually shows when the current item also has video frames. Never
   * releases the player.
   */
  fun setVisible(visible: Boolean) {
    onMain {
      desiredVisible = visible
      updateVisibility()
    }
  }

  /** True when the current item has video frames to present (combined with `desiredVisible`, 2.23). */
  fun hasVideo(): Boolean = currentItemHasVideo

  /**
   * Called from the active target view's `onLayout` so the surface tracks size changes (rotation,
   * mini↔full). Main-thread only.
   */
  fun layoutSurface(view: ViewGroup, target: String) {
    onMain {
      if (activeTarget != target || targetViews[target]?.get() !== view) return@onMain
      val tv = textureView ?: return@onMain
      val params = tv.layoutParams ?: ViewGroup.LayoutParams(view.width, view.height)
      params.width = view.width.coerceAtLeast(1)
      params.height = view.height.coerceAtLeast(1)
      tv.layoutParams = params
    }
  }

  // MARK: - Helpers

  /**
   * Reparent the single texture into the active target's RN view and fill its bounds. Safe no-op
   * when the active target has no registered view yet. Defensive remove-before-add so a concurrent
   * register during modal fragment attach cannot throw `child already has a parent`.
   */
  private fun reparentIntoActiveTarget() {
    val target = activeTarget ?: run {
      updateVisibility()
      return
    }
    val host = targetViews[target]?.get() ?: run {
      updateVisibility()
      return
    }
    val tv = getOrCreateSurface(host.context)
    if (tv.parent === host) {
      updateVisibility()
      return
    }
    (tv.parent as? ViewGroup)?.removeView(tv)
    if (tv.parent != null) {
      (tv.parent as? ViewGroup)?.removeView(tv)
    }
    if (tv.parent == null) {
      val width = if (host.width > 0) host.width else ViewGroup.LayoutParams.MATCH_PARENT
      val height = if (host.height > 0) host.height else ViewGroup.LayoutParams.MATCH_PARENT
      host.addView(tv, ViewGroup.LayoutParams(width, height))
    }
    updateVisibility()
  }

  /** Recompute surface visibility from JS intent + native capability + placement. Main-thread only. */
  private fun updateVisibility() {
    val hasHost = activeTarget?.let { targetViews[it]?.get() } != null
    val visible = desiredVisible && currentItemHasVideo && hasHost
    textureView?.visibility = if (visible) View.VISIBLE else View.GONE
  }

  private fun onMain(work: () -> Unit) {
    if (Looper.myLooper() == Looper.getMainLooper()) work() else mainHandler.post(work)
  }

  private fun <T> onMainSync(block: () -> T): T {
    if (Looper.myLooper() == Looper.getMainLooper()) return block()
    val latch = java.util.concurrent.CountDownLatch(1)
    var result: T? = null
    var error: Throwable? = null
    mainHandler.post {
      try {
        result = block()
      } catch (t: Throwable) {
        error = t
      } finally {
        latch.countDown()
      }
    }
    latch.await()
    error?.let { throw it }
    @Suppress("UNCHECKED_CAST")
    return result as T
  }
}

File: apps/mobile/modules/podverse-media-engine/ios/PodverseVideoSurfaceHost.swift
import AVFoundation
import Foundation
import UIKit

// The ONE native video surface for the process. Mini player and full player are two React views
// over a single engine and a single `AVPlayerLayer`; expanding **reparents** THIS layer's host view
// into the active RN target instead of creating a second `AVPlayer` or mounting a second RN `<Video>`.
//
// The surface is reparented into the RN-mounted `PodverseVideoSurfaceView` for the active target
// (registered from JS), NOT into a process-global window/content overlay. This keeps correct z-order
// and coordinate space in both the base tab view and the modal.

/// Named layout targets the single video surface can occupy. RN mounts a `PodverseVideoSurfaceView`
/// per target and passes the matching `targetId`.
enum PodverseVideoTargetId: String {
  case mini
  case full
}

/// Owns the one `AVPlayerLayer`, bound to `PodverseAudioEngine.shared`'s single player. Access via
/// `PodverseVideoSurfaceHost.shared`.
public final class PodverseVideoSurfaceHost: NSObject {
  /// The one and only video surface host for the process.
  @objc public static let shared = PodverseVideoSurfaceHost()

  /// A `UIView` whose backing layer is the single `AVPlayerLayer`. Using `layerClass` keeps exactly
  /// one layer tied to the view's lifetime — there is never a second video layer.
  final class HostContainerView: UIView {
    override class var layerClass: AnyClass { AVPlayerLayer.self }

    /// The backing `AVPlayerLayer`. Force-cast is the standard idiom for a custom `layerClass`.
    var playerLayer: AVPlayerLayer {
      // swiftlint:disable:next force_cast
      layer as! AVPlayerLayer
    }
  }

  /// Weak reference to an RN target view (so RN unmount deallocates it).
  private final class WeakViewBox {
    weak var view: UIView?
    init(_ view: UIView) { self.view = view }
  }

  /// The transparent overlay view that paints video frames. Reparented into the active target's RN
  /// view; hidden until a video item is active.
  let containerView = HostContainerView()

  /// RN-mounted target views by id. Populated by `PodverseVideoSurfaceView` on attach; the single
  /// surface is reparented into the active target's view.
  private var targetViews: [PodverseVideoTargetId: WeakViewBox] = [:]

  /// The target the surface occupies (`nil` when unplaced / hidden).
  private var activeTarget: PodverseVideoTargetId?

  /// Tracks the last capability reported by the engine. The surface is only shown when the current
  /// item actually has video frames — so a video-medium item playing its audio enclosure never
  /// leaves a black rectangle.
  private var currentItemHasVideo = false

  /// JS-desired visibility for the active target (RN drives this from the playback target kind via
  /// `setVideoSurfaceVisible`). Final visibility = `desiredVisible && currentItemHasVideo`.
  private var desiredVisible = false

  private override init() {
    super.init()
    containerView.isUserInteractionEnabled = false
    containerView.backgroundColor = .clear
    containerView.isHidden = true
    containerView.playerLayer.videoGravity = .resizeAspect
    containerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]

    // Bind the one `AVPlayerLayer` to the one shared `AVPlayer`. This is the only place a layer is
    // attached to the engine's player.
    containerView.playerLayer.player = PodverseAudioEngine.shared.sharedPlayer

    // Record video capability so audio-only items keep the surface hidden.
    PodverseAudioEngine.shared.onVideoCapabilityChanged = { [weak self] hasVideo in
      self?.onMain {
        guard let self = self else { return }
        self.currentItemHasVideo = hasVideo
        self.updateVisibility()
      }
    }
  }

  // MARK: - RN-facing API (called by PodverseVideoSurfaceView + the Expo module)

  /// Register (or replace) the RN view for `target`. The first registered target becomes active so
  /// the single surface is placed; if `target` is the active one, reparent into it immediately.
  /// Never loads/destroys the player.
  func registerTargetView(_ target: PodverseVideoTargetId, view: UIView) {
    onMain { [weak self] in
      guard let self = self else { return }
      self.targetViews[target] = WeakViewBox(view)
      if self.activeTarget == nil {
        self.activeTarget = target
      }
      if self.activeTarget == target {
        self.reparentIntoActiveTarget()
      }
    }
  }

  /// Unregister an RN view (on RN unmount). Only clears when the going-away view is the registered
  /// one (avoids clobbering a re-registered view during a fast remount). If it was the
  /// active target's view, detach the surface (kept alive; just no host).
  func unregisterTargetView(_ target: PodverseVideoTargetId, view: UIView) {
    onMain { [weak self] in
      guard let self = self else { return }
      guard self.targetViews[target]?.view === view else { return }
      self.targetViews[target] = nil
      if self.activeTarget == target {
        self.containerView.removeFromSuperview()
        self.updateVisibility()
      }
    }
  }

  /// Move the single surface to a registered target. Only reparenting/geometry changes; the `AVPlayer`
  /// and playhead are untouched.
  func setActiveTarget(_ target: PodverseVideoTargetId, animatedDuration: TimeInterval = 0) {
    onMain { [weak self] in
      guard let self = self else { return }
      self.activeTarget = target
      self.reparentIntoActiveTarget(animatedDuration: animatedDuration)
    }
  }

  /// Set the JS-desired visibility for the active target (RN drives this from the playback target
  /// kind). The surface only actually shows when the current item also has video frames. Never
  /// tears down the player.
  func setVisible(_ visible: Bool) {
    onMain { [weak self] in
      guard let self = self else { return }
      self.desiredVisible = visible
      self.updateVisibility()
    }
  }

  /// True when the current item has video frames to present. Visibility policy (auto show/hide for
  /// audio-only) combines this with `desiredVisible`.
  func hasVideo() -> Bool { currentItemHasVideo }

  /// Called from the active target view's `layoutSubviews` so the surface tracks size changes
  /// (rotation, split view, full-player expand). Main-thread only.
  func layoutSurface(in view: UIView, for target: PodverseVideoTargetId) {
    onMain { [weak self] in
      guard let self = self else { return }
      guard self.activeTarget == target, self.targetViews[target]?.view === view else { return }
      self.containerView.frame = view.bounds
    }
  }

  // MARK: - Helpers

  /// Reparent the single surface into the active target's RN view and fill its bounds. Safe no-op
  /// (just recomputes visibility) when the active target has no registered view yet — it will
  /// reparent when that view registers.
  private func reparentIntoActiveTarget(animatedDuration: TimeInterval = 0) {
    guard let target = activeTarget, let host = targetViews[target]?.view else {
      updateVisibility()
      return
    }
    if containerView.superview !== host {
      containerView.removeFromSuperview()
      host.addSubview(containerView)
    }
    let apply: () -> Void = { [weak self] in
      guard let self = self else { return }
      self.containerView.frame = host.bounds
    }
    if animatedDuration > 0 {
      UIView.animate(withDuration: animatedDuration, animations: apply)
    } else {
      apply()
    }
    updateVisibility()
  }

  /// Recompute surface visibility from JS intent + native capability + placement. Main-thread only.
  private func updateVisibility() {
    let hasHost = activeTarget.flatMap { targetViews[$0]?.view } != nil
    containerView.isHidden = !(desiredVisible && currentItemHasVideo && hasHost)
  }

  private func onMain(_ work: @escaping () -> Void) {
    if Thread.isMainThread {
      work()
    } else {
      DispatchQueue.main.async(execute: work)
    }
  }
}
